File: src/scripting/server/player.cpp
/************************************************************************************************
 * SwiftlyS2 is a scripting framework for Source2-based games.
 * Copyright (C) 2025 Swiftly Solution SRL via Sava Andrei-Sebastian and it's contributors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 ************************************************************************************************/

#include <api/interfaces/manager.h>
#include <scripting/scripting.h>

#include <api/memory/virtual/call.h>

void Bridge_Player_SendMessage(int playerid, int kind, const char* message, int duration)
{
    static auto playerManager = g_ifaceService.FetchInterface<IPlayerManager>(PLAYERMANAGER_INTERFACE_VERSION);
    auto player = playerManager->GetPlayer(playerid);
    if (!player) return;

    player->SendMsg((MessageType)kind, message, duration);
}

bool Bridge_Player_IsFakeClient(int playerid)
{
    static auto playerManager = g_ifaceService.FetchInterface<IPlayerManager>(PLAYERMANAGER_INTERFACE_VERSION);
    auto player = playerManager->GetPlayer(playerid);
    if (!player) return true;

    return player->IsFakeClient();
}

bool Bridge_Player_IsAuthorized(int playerid)
{
    static auto playerManager = g_ifaceService.FetchInterface<IPlayerManager>(PLAYERMANAGER_INTERFACE_VERSION);
    auto player = playerManager->GetPlayer(playerid);
    if (!player) return false;

    return player->IsAuthorized();
}

uint32_t Bridge_Player_GetConnectedTime(int playerid)
{
    static auto playerManager = g_ifaceService.FetchInterface<IPlayerManager>(PLAYERMANAGER_INTERFACE_VERSION);
    auto player = playerManager->GetPlayer(playerid);
    if (!player) return 0;

    return player->GetConnectedTime();
}

uint64_t Bridge_Player_GetUnauthorizedSteamID(int playerid)
{
    static auto playerManager = g_ifaceService.FetchInterface<IPlayerManager>(PLAYERMANAGER_INTERFACE_VERSION);
    auto player = playerManager->GetPlayer(playerid);
    if (!player) return 0;

    return player->GetUnauthorizedSteamID();
}

uint64_t Bridge_Player_GetSteamID(int playerid)
{
    static auto playerManager = g_ifaceService.FetchInterface<IPlayerManager>(PLAYERMANAGER_INTERFACE_VERSION);
    auto player = playerManager->GetPlayer(playerid);
    if (!player) return 0;

    return player->GetSteamID();
}

void* Bridge_Player_GetController(int playerid)
{
    static auto playerManager = g_ifaceService.FetchInterface<IPlayerManager>(PLAYERMANAGER_INTERFACE_VERSION);
    auto player = playerManager->GetPlayer(playerid);
    if (!player) return nullptr;

    return player->GetController();
}

void* Bridge_Player_GetPawn(int playerid)
{
    static auto playerManager = g_ifaceService.FetchInterface<IPlayerManager>(PLAYERMANAGER_INTERFACE_VERSION);
    auto player = playerManager->GetPlayer(playerid);
    if (!player) return nullptr;

    return player->GetPawn();
}

void* Bridge_Player_GetPlayerPawn(int playerid)
{
    static auto playerManager = g_ifaceService.FetchInterface<IPlayerManager>(PLAYERMANAGER_INTERFACE_VERSION);
    auto player = playerManager->GetPlayer(playerid);
    if (!player) return nullptr;

    return player->GetPlayerPawn();
}

uint64_t Bridge_Player_GetPressedButtons(int playerid)
{
    static auto playerManager = g_ifaceService.FetchInterface<IPlayerManager>(PLAYERMANAGER_INTERFACE_VERSION);
    auto player = playerManager->GetPlayer(playerid);
    if (!player) return 0;

    return player->GetPressedButtons();
}

void Bridge_Player_PerformCommand(int playerid, const char* command)
{
    static auto playerManager = g_ifaceService.FetchInterface<IPlayerManager>(PLAYERMANAGER_INTERFACE_VERSION);
    auto player = playerManager->GetPlayer(playerid);
    if (!player) return;

    player->PerformCommand(command);
}

int Bridge_Player_GetIPAddress(char* out, int playerid)
{
    static auto playerManager = g_ifaceService.FetchInterface<IPlayerManager>(PLAYERMANAGER_INTERFACE_VERSION);
    auto player = playerManager->GetPlayer(playerid);
    if (!player) return 0;

    static std::string s;
    s = player->GetIPAddress();

    if (out != nullptr) strcpy(out, s.c_str());

    return s.size();
}

void Bridge_Player_Kick(int playerid, const char* reason, int gamereason)
{
    static auto playerManager = g_ifaceService.FetchInterface<IPlayerManager>(PLAYERMANAGER_INTERFACE_VERSION);
    auto player = playerManager->GetPlayer(playerid);
    if (!player) return;

    player->Kick(reason, gamereason);
}

void Bridge_Player_ShouldBlockTransmitEntity(int playerid, int entityidx, bool shouldBlockTransmit)
{
    if (playerid + 1 == entityidx) return;

    static auto playerManager = g_ifaceService.FetchInterface<IPlayerManager>(PLAYERMANAGER_INTERFACE_VERSION);
    auto player = playerManager->GetPlayer(playerid);
    if (!player) return;

    auto& bv = player->GetBlockedTransmittingBits();

    auto dword = entityidx / 64;
    if (shouldBlockTransmit) {
        bool wasEmpty = (bv.blockedMask[dword] == 0);
        bv.blockedMask[dword] |= (1 << (entityidx % 64));
        if (wasEmpty) bv.activeMasks.push_back(dword);
    }
    else {
        bv.blockedMask[dword] &= ~(1 << (entityidx % 64));
        if (bv.blockedMask[dword] == 0) bv.activeMasks.erase(std::find(bv.activeMasks.begin(), bv.activeMasks.end(), dword));
    }
}

bool Bridge_Player_IsTransmitEntityBlocked(int playerid, int entityidx)
{
    if (playerid + 1 == entityidx) return false;

    static auto playerManager = g_ifaceService.FetchInterface<IPlayerManager>(PLAYERMANAGER_INTERFACE_VERSION);
    auto player = playerManager->GetPlayer(playerid);
    if (!player) return false;

    auto& bv = player->GetBlockedTransmittingBits();
    return (bv.blockedMask[entityidx / 64] & (1 << (entityidx % 64))) != 0;
}

void Bridge_Player_ClearTransmitEntityBlocked(int playerid)
{
    static auto playerManager = g_ifaceService.FetchInterface<IPlayerManager>(PLAYERMANAGER_INTERFACE_VERSION);
    auto player = playerManager->GetPlayer(playerid);
    if (!player) return;

    auto& bv = player->GetBlockedTransmittingBits();
    for (int i = 0; i < 512; i++) bv.blockedMask[i] = 0;
    bv.activeMasks.clear();
}

void Bridge_Player_ChangeTeam(int playerid, int newteam)
{
    static auto playerManager = g_ifaceService.FetchInterface<IPlayerManager>(PLAYERMANAGER_INTERFACE_VERSION);
    auto player = playerManager->GetPlayer(playerid);
    if (!player) return;

    static auto gamedata = g_ifaceService.FetchInterface<IGameDataManager>(GAMEDATA_INTERFACE_VERSION);
    CALL_VIRTUAL(void, gamedata->GetOffsets()->Fetch("CCSPlayerController::ChangeTeam"), player->GetController(), newteam);
}

void Bridge_Player_SwitchTeam(int playerid, int newteam)
{
    static auto playerManager = g_ifaceService.FetchInterface<IPlayerManager>(PLAYERMANAGER_INTERFACE_VERSION);
    auto player = playerManager->GetPlayer(playerid);
    if (!player) return;

    static auto gamedata = g_ifaceService.FetchInterface<IGameDataManager>(GAMEDATA_INTERFACE_VERSION);
    if (newteam == 0 || newteam == 1)
        CALL_VIRTUAL(void, gamedata->GetOffsets()->Fetch("CCSPlayerController::ChangeTeam"), player->GetController(), newteam);
    else
        reinterpret_cast<void(*)(void*, int)>(gamedata->GetSignatures()->Fetch("CCSPlayerController::SwitchTeam"))(player->GetController(), newteam);
}

void Bridge_Player_TakeDamage(int playerid, void* dmginfo)
{
    static auto playerManager = g_ifaceService.FetchInterface<IPlayerManager>(PLAYERMANAGER_INTERFACE_VERSION);
    auto player = playerManager->GetPlayer(playerid);
    if (!player) return;

    static auto gamedata = g_ifaceService.FetchInterface<IGameDataManager>(GAMEDATA_INTERFACE_VERSION);
    reinterpret_cast<int64_t(*)(void*, void*, void*)>(gamedata->GetSignatures()->Fetch("CBaseEntity::TakeDamage"))(player->GetPawn(), dmginfo, 0);
}

void Bridge_Player_Teleport(int playerid, Vector pos, QAngle angle, Vector vel)
{
    static auto playerManager = g_ifaceService.FetchInterface<IPlayerManager>(PLAYERMANAGER_INTERFACE_VERSION);
    auto player = playerManager->GetPlayer(playerid);
    if (!player) return;

    static auto gamedata = g_ifaceService.FetchInterface<IGameDataManager>(GAMEDATA_INTERFACE_VERSION);
    CALL_VIRTUAL(void, gamedata->GetOffsets()->Fetch("CBaseEntity::Teleport"), player->GetPawn(), &pos, &angle, &vel);
}

int Bridge_Player_GetLanguage(char* out, int playerid)
{
    static auto playerManager = g_ifaceService.FetchInterface<IPlayerManager>(PLAYERMANAGER_INTERFACE_VERSION);
    auto player = playerManager->GetPlayer(playerid);
    if (!player) return 0;

    static std::string s;
    s = player->GetLanguage();

    if (out != nullptr) strcpy(out, s.c_str());

    return s.size();
}

void Bridge_Player_SetCenterMenuRender(int playerid, const char* text)
{
    static auto playerManager = g_ifaceService.FetchInterface<IPlayerManager>(PLAYERMANAGER_INTERFACE_VERSION);
    auto player = playerManager->GetPlayer(playerid);
    if (!player) return;

    player->RenderMenuCenterText(text);
}

void Bridge_Player_ClearCenterMenuRender(int playerid)
{
    static auto playerManager = g_ifaceService.FetchInterface<IPlayerManager>(PLAYERMANAGER_INTERFACE_VERSION);
    auto player = playerManager->GetPlayer(playerid);
    if (!player) return;

    player->ClearRenderMenuCenterText();
}

bool Bridge_Player_HasMenuShown(int playerid)
{
    static auto playerManager = g_ifaceService.FetchInterface<IPlayerManager>(PLAYERMANAGER_INTERFACE_VERSION);
    auto player = playerManager->GetPlayer(playerid);
    if (!player) return false;

    return player->HasMenuShown();
}

void Bridge_Player_ExecuteCommand(int playerid, const char* command)
{
    static auto playerManager = g_ifaceService.FetchInterface<IPlayerManager>(PLAYERMANAGER_INTERFACE_VERSION);
    auto player = playerManager->GetPlayer(playerid);
    if (!player) return;

    CCommand cmd;
    cmd.Tokenize(command);

    ConCommandRef cmdRef(cmd[0]);

    if (cmdRef.IsValidRef()) {
        CCommandContext context(CommandTarget_t::CT_FIRST_SPLITSCREEN_CLIENT, CPlayerSlot(player->GetSlot()));
        cmdRef.Dispatch(context, cmd);
    }
    else {
        static auto engine = g_ifaceService.FetchInterface<IVEngineServer2>(INTERFACEVERSION_VENGINESERVER);
        engine->ClientCommand(player->GetSlot(), command);
    }
}

DEFINE_NATIVE("Player.SendMessage", Bridge_Player_SendMessage);
DEFINE_NATIVE("Player.IsFakeClient", Bridge_Player_IsFakeClient);
DEFINE_NATIVE("Player.IsAuthorized", Bridge_Player_IsAuthorized);
DEFINE_NATIVE("Player.GetConnectedTime", Bridge_Player_GetConnectedTime);
DEFINE_NATIVE("Player.GetUnauthorizedSteamID", Bridge_Player_GetUnauthorizedSteamID);
DEFINE_NATIVE("Player.GetSteamID", Bridge_Player_GetSteamID);
DEFINE_NATIVE("Player.GetController", Bridge_Player_GetController);
DEFINE_NATIVE("Player.GetPawn", Bridge_Player_GetPawn);
DEFINE_NATIVE("Player.GetPlayerPawn", Bridge_Player_GetPlayerPawn);
DEFINE_NATIVE("Player.GetPressedButtons", Bridge_Player_GetPressedButtons);
DEFINE_NATIVE("Player.PerformCommand", Bridge_Player_PerformCommand);
DEFINE_NATIVE("Player.GetIPAddress", Bridge_Player_GetIPAddress);
DEFINE_NATIVE("Player.Kick", Bridge_Player_Kick);
DEFINE_NATIVE("Player.ShouldBlockTransmitEntity", Bridge_Player_ShouldBlockTransmitEntity);
DEFINE_NATIVE("Player.IsTransmitEntityBlocked", Bridge_Player_IsTransmitEntityBlocked);
DEFINE_NATIVE("Player.ClearTransmitEntityBlocked", Bridge_Player_ClearTransmitEntityBlocked);
DEFINE_NATIVE("Player.ChangeTeam", Bridge_Player_ChangeTeam);
DEFINE_NATIVE("Player.SwitchTeam", Bridge_Player_SwitchTeam);
DEFINE_NATIVE("Player.TakeDamage", Bridge_Player_TakeDamage);
DEFINE_NATIVE("Player.Teleport", Bridge_Player_Teleport);
DEFINE_NATIVE("Player.GetLanguage", Bridge_Player_GetLanguage);
DEFINE_NATIVE("Player.SetCenterMenuRender", Bridge_Player_SetCenterMenuRender);
DEFINE_NATIVE("Player.ClearCenterMenuRender", Bridge_Player_ClearCenterMenuRender);
DEFINE_NATIVE("Player.HasMenuShown", Bridge_Player_HasMenuShown);
DEFINE_NATIVE("Player.ExecuteCommand", Bridge_Player_ExecuteCommand);

File: managed/src/SwiftlyS2.Generated/Natives/Player.cs
#pragma warning disable CS0649
#pragma warning disable CS0169

using System.Buffers;
using System.Text;
using System.Threading;
using SwiftlyS2.Shared.Natives;

namespace SwiftlyS2.Core.Natives;

internal static class NativePlayer {

  private unsafe static delegate* unmanaged<int, int, byte*, int, void> _SendMessage;

  public unsafe static void SendMessage(int playerid, int kind, string message, int htmlDuration) {
    if (!NativeBinding.IsMainThread) {
      throw new InvalidOperationException("This method can only be called from the main thread.");
    }
    var pool = ArrayPool<byte>.Shared;
    var messageLength = Encoding.UTF8.GetByteCount(message);
    var messageBuffer = pool.Rent(messageLength + 1);
    Encoding.UTF8.GetBytes(message, messageBuffer);
    messageBuffer[messageLength] = 0;
    fixed (byte* messageBufferPtr = messageBuffer) {
      _SendMessage(playerid, kind, messageBufferPtr, htmlDuration);
      pool.Return(messageBuffer);
    }
  }

  private unsafe static delegate* unmanaged<int, byte> _IsFakeClient;

  public unsafe static bool IsFakeClient(int playerid) {
    var ret = _IsFakeClient(playerid);
    return ret == 1;
  }

  private unsafe static delegate* unmanaged<int, byte> _IsAuthorized;

  public unsafe static bool IsAuthorized(int playerid) {
    var ret = _IsAuthorized(playerid);
    return ret == 1;
  }

  private unsafe static delegate* unmanaged<int, uint> _GetConnectedTime;

  public unsafe static uint GetConnectedTime(int playerid) {
    var ret = _GetConnectedTime(playerid);
    return ret;
  }

  private unsafe static delegate* unmanaged<int, ulong> _GetUnauthorizedSteamID;

  public unsafe static ulong GetUnauthorizedSteamID(int playerid) {
    var ret = _GetUnauthorizedSteamID(playerid);
    return ret;
  }

  private unsafe static delegate* unmanaged<int, ulong> _GetSteamID;

  public unsafe static ulong GetSteamID(int playerid) {
    var ret = _GetSteamID(playerid);
    return ret;
  }

  private unsafe static delegate* unmanaged<int, nint> _GetController;

  public unsafe static nint GetController(int playerid) {
    var ret = _GetController(playerid);
    return ret;
  }

  private unsafe static delegate* unmanaged<int, nint> _GetPawn;

  public unsafe static nint GetPawn(int playerid) {
    var ret = _GetPawn(playerid);
    return ret;
  }

  private unsafe static delegate* unmanaged<int, nint> _GetPlayerPawn;

  public unsafe static nint GetPlayerPawn(int playerid) {
    var ret = _GetPlayerPawn(playerid);
    return ret;
  }

  private unsafe static delegate* unmanaged<int, ulong> _GetPressedButtons;

  public unsafe static ulong GetPressedButtons(int playerid) {
    var ret = _GetPressedButtons(playerid);
    return ret;
  }

  private unsafe static delegate* unmanaged<int, byte*, void> _PerformCommand;

  public unsafe static void PerformCommand(int playerid, string command) {
    if (!NativeBinding.IsMainThread) {
      throw new InvalidOperationException("This method can only be called from the main thread.");
    }
    var pool = ArrayPool<byte>.Shared;
    var commandLength = Encoding.UTF8.GetByteCount(command);
    var commandBuffer = pool.Rent(commandLength + 1);
    Encoding.UTF8.GetBytes(command, commandBuffer);
    commandBuffer[commandLength] = 0;
    fixed (byte* commandBufferPtr = commandBuffer) {
      _PerformCommand(playerid, commandBufferPtr);
      pool.Return(commandBuffer);
    }
  }

  private unsafe static delegate* unmanaged<byte*, int, int> _GetIPAddress;

  public unsafe static string GetIPAddress(int playerid) {
    var ret = _GetIPAddress(null, playerid);
    var pool = ArrayPool<byte>.Shared;
    var retBuffer = pool.Rent(ret + 1);
    fixed (byte* retBufferPtr = retBuffer) {
      ret = _GetIPAddress(retBufferPtr, playerid);
      var retString = Encoding.UTF8.GetString(retBufferPtr, ret);
      pool.Return(retBuffer);
      return retString;
    }
  }

  private unsafe static delegate* unmanaged<int, byte*, int, void> _Kick;

  public unsafe static void Kick(int playerid, string reason, int gamereason) {
    if (!NativeBinding.IsMainThread) {
      throw new InvalidOperationException("This method can only be called from the main thread.");
    }
    var pool = ArrayPool<byte>.Shared;
    var reasonLength = Encoding.UTF8.GetByteCount(reason);
    var reasonBuffer = pool.Rent(reasonLength + 1);
    Encoding.UTF8.GetBytes(reason, reasonBuffer);
    reasonBuffer[reasonLength] = 0;
    fixed (byte* reasonBufferPtr = reasonBuffer) {
      _Kick(playerid, reasonBufferPtr, gamereason);
      pool.Return(reasonBuffer);
    }
  }

  private unsafe static delegate* unmanaged<int, int, byte, void> _ShouldBlockTransmitEntity;

  public unsafe static void ShouldBlockTransmitEntity(int playerid, int entityidx, bool shouldBlockTransmit) {
    _ShouldBlockTransmitEntity(playerid, entityidx, shouldBlockTransmit ? (byte)1 : (byte)0);
  }

  private unsafe static delegate* unmanaged<int, int, byte> _IsTransmitEntityBlocked;

  public unsafe static bool IsTransmitEntityBlocked(int playerid, int entityidx) {
    var ret = _IsTransmitEntityBlocked(playerid, entityidx);
    return ret == 1;
  }

  private unsafe static delegate* unmanaged<int, void> _ClearTransmitEntityBlocked;

  public unsafe static void ClearTransmitEntityBlocked(int playerid) {
    _ClearTransmitEntityBlocked(playerid);
  }

  private unsafe static delegate* unmanaged<int, int, void> _ChangeTeam;

  public unsafe static void ChangeTeam(int playerid, int newteam) {
    if (!NativeBinding.IsMainThread) {
      throw new InvalidOperationException("This method can only be called from the main thread.");
    }
    _ChangeTeam(playerid, newteam);
  }

  private unsafe static delegate* unmanaged<int, int, void> _SwitchTeam;

  public unsafe static void SwitchTeam(int playerid, int newteam) {
    if (!NativeBinding.IsMainThread) {
      throw new InvalidOperationException("This method can only be called from the main thread.");
    }
    _SwitchTeam(playerid, newteam);
  }

  private unsafe static delegate* unmanaged<int, nint, void> _TakeDamage;

  public unsafe static void TakeDamage(int playerid, nint dmginfo) {
    if (!NativeBinding.IsMainThread) {
      throw new InvalidOperationException("This method can only be called from the main thread.");
    }
    _TakeDamage(playerid, dmginfo);
  }

  private unsafe static delegate* unmanaged<int, Vector, QAngle, Vector, void> _Teleport;

  public unsafe static void Teleport(int playerid, Vector pos, QAngle angle, Vector velocity) {
    if (!NativeBinding.IsMainThread) {
      throw new InvalidOperationException("This method can only be called from the main thread.");
    }
    _Teleport(playerid, pos, angle, velocity);
  }

  private unsafe static delegate* unmanaged<byte*, int, int> _GetLanguage;

  public unsafe static string GetLanguage(int playerid) {
    var ret = _GetLanguage(null, playerid);
    var pool = ArrayPool<byte>.Shared;
    var retBuffer = pool.Rent(ret + 1);
    fixed (byte* retBufferPtr = retBuffer) {
      ret = _GetLanguage(retBufferPtr, playerid);
      var retString = Encoding.UTF8.GetString(retBufferPtr, ret);
      pool.Return(retBuffer);
      return retString;
    }
  }

  private unsafe static delegate* unmanaged<int, byte*, void> _SetCenterMenuRender;

  public unsafe static void SetCenterMenuRender(int playerid, string text) {
    var pool = ArrayPool<byte>.Shared;
    var textLength = Encoding.UTF8.GetByteCount(text);
    var textBuffer = pool.Rent(textLength + 1);
    Encoding.UTF8.GetBytes(text, textBuffer);
    textBuffer[textLength] = 0;
    fixed (byte* textBufferPtr = textBuffer) {
      _SetCenterMenuRender(playerid, textBufferPtr);
      pool.Return(textBuffer);
    }
  }

  private unsafe static delegate* unmanaged<int, void> _ClearCenterMenuRender;

  public unsafe static void ClearCenterMenuRender(int playerid) {
    _ClearCenterMenuRender(playerid);
  }

  private unsafe static delegate* unmanaged<int, byte> _HasMenuShown;

  public unsafe static bool HasMenuShown(int playerid) {
    var ret = _HasMenuShown(playerid);
    return ret == 1;
  }

  private unsafe static delegate* unmanaged<int, byte*, void> _ExecuteCommand;

  public unsafe static void ExecuteCommand(int playerid, string command) {
    if (!NativeBinding.IsMainThread) {
      throw new InvalidOperationException("This method can only be called from the main thread.");
    }
    var pool = ArrayPool<byte>.Shared;
    var commandLength = Encoding.UTF8.GetByteCount(command);
    var commandBuffer = pool.Rent(commandLength + 1);
    Encoding.UTF8.GetBytes(command, commandBuffer);
    commandBuffer[commandLength] = 0;
    fixed (byte* commandBufferPtr = commandBuffer) {
      _ExecuteCommand(playerid, commandBufferPtr);
      pool.Return(commandBuffer);
    }
  }

  private unsafe static delegate* unmanaged<int, byte> _IsFirstSpawn;

  public unsafe static bool IsFirstSpawn(int playerid) {
    var ret = _IsFirstSpawn(playerid);
    return ret == 1;
  }

  private unsafe static delegate* unmanaged<int, int> _GetUserID;

  public unsafe static int GetUserID(int playerid) {
    var ret = _GetUserID(playerid);
    return ret;
  }
}